packages/deployments/contracts/contracts/nomad-xapps/contracts/promise-router/PromiseRouter.sol

```diff
@@ -154,23 +211,58 @@ contract PromiseRouter is Version0, Router {
 
     bytes32 transferId = _msg.transferId();
     address callbackAddress = _msg.callbackAddress();
-    bytes memory data = _msg.returnCallData();
+    bool success = _msg.returnSuccess();
+    bytes memory data = _msg.returnData();
 
-    //TODO process callback
+    //store Promise message
+    promiseMessages[transferId] = _message;
 
     // emit Receive event
-    emit Receive(_originAndNonce(_origin, _nonce), _origin, transferId, callbackAddress, data);
+    emit Receive(_originAndNonce(_origin, _nonce), _origin, transferId, callbackAddress, success, data, _message);
   }
 
-  function process(bytes32 transactionId, bytes memory _message) internal {
-    // Should parse out the return data and callback address from message
-    // Should execute the callback() function on the provided Callback address
-    // Should enforce relayer is whitelisted by calling local connext contract
+  function process(bytes32 transferId) public payable nonReentrant {
+    // parse out the return data and callback address from message
+    bytes memory _message = promiseMessages[transferId];
+    if (_message.length == 0) revert PromiseRouter__process_InvalidMessage();
+
+    bytes29 _msg = _message.ref(0).mustBePromiseCallback();
+
+    // enforce relayer is whitelisted by calling local connext contract
+    if (!connext.isApprovedRelayer(msg.sender)) revert PromiseRouter__process_notApprovedRelayer();
+
+    // enfore callback fee is not zero
+    if (callbackFees[transferId] == 0) revert PromiseRouter__process_insufficientCallbackFee();
```

Comment:
What happens if the callback executor doesn't want a fee (i.e. what if it's the user)?

Reply:
Idk if the optimizer is handling this, but we can probs afford to assign to memory a value here:
```
uint256 fee = callbackFees[transferId];
```
then we don't have to do a second SLOAD/read operation below